### DISCUSSION TITLE
Laravel ULID support

### DISCUSSION THREAD

**Discussion by guifelix** [laravel/framework#37029] [Ideas]
I wonder what Laravel core team think about having [ULID](https://github.com/ulid/spec) on the framework since it has many advantages when comparing with UUID.

If this is a good idea, what would be needed to do? 

1. add it on `src/Illuminate/Support/Str.php` like [here](https://github.com/laravel/framework/blob/02fd52f65491e45a768068e3a6c731fdf2d35ac1/src/Illuminate/Support/Str.php#L291-L304) and [here](https://github.com/laravel/framework/blob/02fd52f65491e45a768068e3a6c731fdf2d35ac1/src/Illuminate/Support/Str.php#L780-L836)?
2. create a new separate package to add the functionality?

E.g. (made with [this port of ULID](https://github.com/robinvdvleuten/php-ulid) in mind)
```php
<?php

namespace Illuminate\Support;

use Illuminate\Support\Traits\Macroable;
use League\CommonMark\GithubFlavoredMarkdownConverter;
use Ramsey\Uuid\Codec\TimestampFirstCombCodec;
use Ramsey\Uuid\Generator\CombGenerator;
use Ramsey\Uuid\Uuid;
use Ramsey\Uuid\UuidFactory;
use voku\helper\ASCII;
use Ulid\Ulid;

class Str
{
...
    /**
     * Determine if a given string is a valid ULID.
     *
     * @param string $value
     * @return bool
     */
    public static function isUlid($value)
    {
        if (!is_string($value)) {
            return false;
        }

        return preg_match('/^([0-9]|[ABCDEFGHJKMNPQRSTVWXYZ]){26}$/iD', $value) > 0;
    }
...
    /**
     * Generate a ULID
     *
     * @param bool $lowercase 
     * @return \Ulid\Ulid
     */
    public static function ulid(bool $lowercase = false)
    {
        return Ulid::generate($lowercase);
    }

    /**
     * Return timestamp from an ULID
     *
     * @param \Ulid\Ulid $subject
     * @return int
     */
    public static function ulidTimestamp(Ulid $subject)
    {
        return $subject->toTimestamp();
    }

    /**
     * Generate a ULID from a timestamp
     *
     * @param int  $value
     * @param bool $lowercase
     * @return \Ulid\Ulid
     */
    public static function ulidFromTimestamp(int $value, bool $lowercase = false)
    {
        return Ulid::fromTimestamp($value, $lowercase);
    }

    /**
     * Generate a ULID from a string
     *
     * @param int  $value
     * @param bool $lowercase
     * @return \Ulid\Ulid
     */
    public static function ulidFromString(string $value, bool $lowercase = false)
    {
        return Ulid::fromString($value, $lowercase);
    }

    /**
     * Return time from an ULID
     *
     * @param \Ulid\Ulid $subject
     * @return string
     */
    public static function ulidTime(Ulid $subject)
    {
        return $subject->getTime();
    }

    /**
     * Return Randomness from an ULID
     *
     * @param \Ulid\Ulid $subject
     * @return string
     */
    public static function ulidRandomness(Ulid $subject)
    {
        return $subject->getRandomness();
    }
```

**Comment by devfrey** [comment]
> since it has many advantages when comparing with UUID

What are these advantages, realistically, in a Laravel project? What's your use case?

If you want UUIDs that are sortable, take a look at [Str::orderedUuid()](https://laravel.com/docs/8.x/helpers#method-str-ordered-uuid).

**Reply by guifelix** [reply]
I don't specifically have a use case myself at the moment, I thought of adding ULIDs in case anyone want to use them as UUIDs are available on Laravel as well, I don't know if I'd add this feature to Laravel core just to solve my specific problem, hence the original question of what the core think about it and whether they think it should be a core feature or a separate project. But if you'd asked me if I'd use ULIDs over UUIDs if it was available, the answer would be yes, and maybe other people would as well, maybe other people use UUIDs just because they don't know ULIDs as Laravel introduced UUIDs to them and they don't think there's another option.

In a Laravel project, the advantages of using ULIDs are the same as having UUIDs over IDs, and the advantages over UUIDs specifically, they're stated on ULID's project:

> UUID can be suboptimal for many use-cases because:
>
>- It isn't the most character efficient way of encoding 128 bits of randomness
>- UUID v1/v2 is impractical in many environments, as it requires access to a unique, stable MAC address
>- UUID v3/v5 requires a unique seed and produces randomly distributed IDs, which can cause fragmentation in many data structures
>- UUID v4 provides no other information than randomness which can cause fragmentation in many data structures

**Comment by crynobone** [comment]
This should already be supported.
